File: C++Verifier/src/expressions/ArithExpr.cpp
#ifndef ARITHEXPR_H
#define ARITHEXPR_H

#include <vector>
#include "Expr.h"

using namespace std;

class ArithExpr: public virtual Expr {
public:
    using Expr::Expr;

    virtual void* evaluate(shared_ptr<Packet> p) = 0;

    virtual string evaluate(shared_ptr<Packet> p,  vector<string> constraints, vector<string> locationList, vector<string> variableList) = 0;

    static string applyArithOp(string leftValue, string rightValue, string operand) {
        long leftVal = stol(leftValue);
        long rightVal = stol(rightValue);
        if (operand == "+") {
            return to_string(leftVal + rightVal);
        } else if (operand == "-") {
            return to_string(leftVal - rightVal);
        } else if (operand == "*") {
            return to_string(leftVal * rightVal);
        } else if (operand == "/") {
            return to_string(leftVal / rightVal);
        }
        throw std::runtime_error("ArithExpr operand not supported");

        return "0";
    }
    virtual ~ArithExpr() {};

};

#endif

File: verification/src/main/java/org/msr/mnr/verification/expressions/ArithExpr.java
package org.msr.mnr.verification.expressions;

import java.util.ArrayList;

import org.msr.mnr.verification.utils.Packet;
import org.msr.mnr.verification.utils.ParseIntArray;

public abstract class ArithExpr extends Expr {
    private static final long serialVersionUID = 202005082232L;

    public ArithExpr(Object left, Object right) {
        super(left, right);
    }

    public ArithExpr(Object left) {
        super(left);
    }

    @Override
    public abstract int[] evaluate(Packet p);

    public abstract int[] evaluate(Packet p, ArrayList<int[]> constraints,
            ArrayList<String> locationList, ArrayList<String> variableList);

    public static int[] applyArithOp(int[] leftValue, int[] rightValue, String operand) {
        long leftVal = ParseIntArray.getLong(leftValue);
        long rightVal = ParseIntArray.getLong(rightValue);
        switch(operand) {
        case "+":
            return ParseIntArray.fromLong(leftVal + rightVal);
        case "-":
            return ParseIntArray.fromLong(leftVal - rightVal);
        case "*":
            return ParseIntArray.fromLong(leftVal * rightVal);
        case "/":
            return ParseIntArray.fromLong(leftVal / rightVal);
        default:
            break;
        }
        throw new RuntimeException("Arith Operation not found: " + operand);
    }

}